# The functions required to perform the simulation study for my X chromosome association testing in GWAS paper# Peter Hickey 23/12/2010# UPDATE 21/12/2010# Added a 4df chi-square test of the combined male and female genotype table# Currently simulating under HWE for the X chromosome as defined by Lange and using my way to calculate genotype frequencies separately in males and females#library(Hmisc)# ABT(t) (allele based test) has a N(0,1) distribution under the null# ABT(t)^2 is equivalent to chisq.test(t, correct=F)ABT <- function(t){# X_A.2 is Z_ABT^2#X_A.2 <- (sum(t)*(sum(t)*t[1,2] - sum(t[1,])*sum(t[,2]))^2)/(sum(t[1,])*sum(t[2,])*(sum(t)*(sum(t[,2])) - (sum(t[,2]))^2))Z_ABT <- (sqrt(sum(t))*(sum(t)*t[1,2] - sum(t[1,])*sum(t[,2])))/sqrt((sum(t[1,])*sum(t[2,])*(sum(t)*(sum(t[,2])) - (sum(t[,2]))^2)))#return(1-pchisq(X_A.2,1))return(Z_ABT)}# Z_mfA test as defined by Zheng (2007). Has an asymptotic N(0,1) distribution under the nullZ_mfA <- function(females, males){females.allele <- matrix(c(2*females[1,1]+females[1,2],2*females[1,3]+females[1,2],2*females[2,1]+females[2,2],2*females[2,3]+females[2,2]),2,2,byrow=T)Z_m <- ABT(males)Z_fA <- ABT(females.allele)Z_mfA.stat <- (sqrt(sum(males)/(sum(males)+sum(females.allele))))*Z_m + (sqrt(sum(females.allele)/(sum(males)+sum(females.allele))))*Z_fAreturn(Z_mfA.stat)}# Z_mfG test as defined by Zheng (2007). Has an asymptotic N(0,1) distribution under the nullZ_mfG <- function(females,males){Z_m <- ABT(males)Z_fG <- trendTest(females)Z_mfg.stat <- (sqrt(sum(males)/(sum(males)+sum(females))))*Z_m + (sqrt(sum(females)/(sum(males)+sum(females))))*Z_fGreturn(Z_mfg.stat)}# trendTest (Cochran-Armitage trend test for genotype tables) has a N(0,1) distribution under the nulltrendTest <- function(t){N <- sum(t)r1 <- t[1,2]r2 <- t[1,3]R <- sum(t[1,])n1 <- sum(t[,2])n2 <- sum(t[,3])Z_G <- (sqrt(N)*(N*(r1+2*r2)-R*(n1+2*n2)))/sqrt(R*(N-R)*(N*(n1+4*n2)-(n1+2*n2)^2))return(Z_G)}# The allele based test applied across the combined male and female samplesZ_A.combined <- function(females,males){t.combined <- matrix(c(2*females[1,1]+females[1,2]+males[1,1], 2*females[1,3]+females[1,2]+males[1,2],2*females[2,1]+females[2,2]+males[2,1], 2*females[2,3]+females[2,2]+males[2,2]),2,2,byrow=T)Z_A <- ABT(t.combined)return(Z_A)}# Added 21/12/2010# A 4 degree of freedom Pearson chi-square test of the combined male and female genotype table# Table is of the form shown below#                       AA      AB      BB      A- B-# Case# ControlchiComb <- function(females,males){        t.combined2 <- cbind(females,males) # This must NOT be t.combined as this variable already used in ABT()        return(chisq.test(t.combined2)$statistic)        }# 20/07/2010: A re-write of claytonTestSimulation claytonTestSimulation <- function(ped){ped.male <- subset(ped, SEX == 1)ped.female <- subset(ped, SEX == 2)A.m <- ped.male$GENOTYPEA.f <- ped.female$GENOTYPEY.m <- ped.male$PHENOTYPEY.f <- ped.female$PHENOTYPED.f <- ped.female$DOM #  A dominance indicator. If female is het D=1, else D=0D.m <- rep(0,nrow(ped.male)) # A vector of dominance term indicators in males. Identically 0 as there are no male hets for non-par X SNPsY <- c(Y.f, Y.m) # A vector of all phenotypes with females firstA <- c(A.f, A.m) # A vector of all genotypes with females firstD <- c(D.f, D.m) # A vector of all dominance indicator terms with females firstN <- length(Y)U_A <- sum((Y-mean(Y))*A) # Score for AF <- nrow(ped.female) # Number of females U_D <- sum((Y.f - mean(Y.f))*D.f) # Score for DU <- matrix(c(U_A,U_D),2,1)# Now calculate the covariance matrices     # We use mean(A) rather than mean(A.f) as we are assuming that allele frequencies to be equal# between males and femalesVhat_F.11 <- sum((A.f-mean(A))^2)Vhat_F.12 <- sum((A.f-mean(A))*(D.f-mean(D.f)))Vhat_F.21 <- Vhat_F.12Vhat_F.22 <- sum((D.f-mean(D.f))^2)Vhat_F <- (1/(F-1)) * matrix(c(Vhat_F.11, Vhat_F.12, Vhat_F.21, Vhat_F.22), 2,2, byrow=T)Palt <- mean(A,na.rm=T)/2Vhat_M <- matrix(c(4*Palt*(1-Palt), 0, 0, 0), 2, 2, byrow=T) Vhat <- Vhat_F * sum((Y.f-mean(Y))^2) + Vhat_M * sum((Y.m-mean(Y))^2)# Now to calculate the test statistic# NB: Matrix multiplication operator is %*% and matrix inverse operator is "solve"scoreStat2df <- t(U) %*% solve(Vhat) %*% UscoreStat1df <- (U_A)^2/Vhat[1,1]results <- data.frame(N = N,Chi.squared.1.df = scoreStat1df,Chi.squared.2.df = scoreStat2df,P.1df = 1 - pchisq(scoreStat1df,1),P.2df = 1 - pchisq(scoreStat2df,2))return(results)}##################################################################################### Compute the various statistics for the same n_reps simulations##################################################################################### XSim simulates X chromosome genotypes and calculates several different test statistics for testing associations of X chromosome genotype and phenotypeXSim <- function(n_reps,N_cases,N_controls,case.sex.ratio,control.sex.ratio,MAF,MODEL,t,K,results=NULL,alpha=0.05){# case.sex.ratio = Number of female cases/Total number of cases# control.sex.ratio = Number of female controls/Total number of controls# K is the disease prevalence; K=sum(f*g) Freidman (2002)N <- N_cases + N_controlsN_cases.f <- N_cases*case.sex.ratioN_cases.m <- N_cases - N_cases.fN_controls.f <- N_controls*control.sex.ratioN_controls.m <- N_controls - N_controls.fN_f <- N_cases.f + N_controls.fN_m <- N_cases.m + N_controls.m# Two possible alleles: A and B# We consider B to be the risk allele, thus MAF=Pr(B) (in general population)# Specify the Genotypic Relative Risk (GRR) ratios # gamma = (gamma_0,gamma_1, gamma_2) as defined by Zheng (2007), Schaid (2001), Freidlin (2002)# gamma_1 = Pr(case|AB)/Pr(case|AA), gamma_2 = Pr(case|BB)/Pr(case|AA)# NB: gamma_0 = 1 by definition; see Freidlin (2002)if(MODEL=="Null"){gamma <- c(1,1,1)}if(MODEL=="REC"){gamma <- c(1,1,t)}if(MODEL=="ADD"){gamma <- c(1,(1+t)/2, t)}if(MODEL=="DOM"){gamma <- c(1,t, t)}# g.f=(g.f_0,g.f_1,g.f_2)=(g.f_AA,g.f_AB,g.f_BB) is the vector of genotype frequencies in the wider FEMALE population# g.m=(g.m_0,g.m_2)=(g.m_A-,g.m_B-) is the vector of genotype frequencies in the wider MALE population# NB: Currently assuming HWEg.f <- c((1-MAF)^2,2*MAF*(1-MAF),MAF^2)g.m <- c(1-MAF, MAF)# f=(f_0,f_1,f_2) is the vector of (non-standardized) penetrances for the genotypes AA,AB,BB respectivelyf.f_0 <- K/(g.f[3]*gamma[3]+g.f[2]*gamma[2]+g.f[1]) f.f <- c(f.f_0,gamma[2]*f.f_0,gamma[3]*f.f_0)f.m <- c(f.f[1], f.f[3]) # We assume male hemizygous penetrances are equal to female homozygous penetrances# p.f=(p.f_0,p.f_1,p.f_2) is the vector of genotype frequencies in the FEMALE cases# q.f=(q.f_0,q.f_1,q.f_2) is the vector of genotype frequencies in the FEMALE controlsp.f <- (f.f*g.f)/sum(f.f*g.f)q.f <- ((1-f.f)*g.f)/sum((1-f.f)*g.f)# p.m=(p.m_0,p.m_2) is the vector of genotype frequencies in the MALE cases# q.m=(q.m_0,q.m_2) is the vector of genotype frequencies in the MALE controlsp.m <- (f.m*g.m)/sum(f.m*g.m)q.m <- ((1-f.m)*g.m)/sum((1-f.m)*g.m)# Vectors to store values of each test statisticsZ_A.sq <- rep(0,n_reps)Z_C.sq <- rep(0,n_reps)Z_mfA.sq <- rep(0,n_reps)Z_mfG.sq <- rep(0,n_reps)clayton1df <- rep(0,n_reps)clayton2df <- rep(0,n_reps)plink.trend.f <- rep(0,n_reps)plink.geno.f <- rep(0,n_reps)chi4df <- rep(0,n_reps)for(k in 1:n_reps){A_cases.f <- as.vector(t(rmultinom(1,N_cases.f,p.f)))A_controls.f <- as.vector(t(rmultinom(1,N_controls.f,q.f)))t.f <- matrix(c(A_cases.f,A_controls.f),2,3,byrow=T)A_cases.m <- as.vector(t(rmultinom(1,N_cases.m,p.m)))A_controls.m <- as.vector(t(rmultinom(1,N_controls.m,q.m)))t.m <- matrix(c(A_cases.m,A_controls.m),2,2,byrow=T)# Turn the genotype tables into individual data - similar to a pedigree filepheno.f <- c(rep(1,sum(t.f[1,])),rep(0,sum(t.f[2,])))geno.f <- NULLfor (i in 1:2){        for(j in 1:3){        geno.f <- append(geno.f, rep(j-1,t.f[i,j]))        }}pheno.m <- c(rep(1,sum(t.m[1,])),rep(0,sum(t.m[2,])))geno.m <- NULLfor (i in 1:2){        for(j in 1:2){        if(j==1){        geno.m <- append(geno.m, rep(0,t.m[i,j]))        }         else{                geno.m <- append(geno.m,rep(2,t.m[i,j]))                }        }}pheno <- c(pheno.f,pheno.m)geno <- c(geno.f,geno.m)sex <- c(rep(2,length(pheno.f)), rep(1,length(pheno.m)))ped <- data.frame(PHENOTYPE = pheno, GENOTYPE = geno, SEX = sex)ped$DOM <- ifelse(ped$SEX==2&ped$GENOTYPE==1,1,0)Z_A.sq[k] <- (Z_A.combined(t.f,t.m))^2Z_m <- ABT(t.m)Z_fG <- trendTest(t.f)Z_C.sq[k] <- (Z_m)^2 + (Z_fG)^2Z_mfA.sq[k] <- (Z_mfA(t.f,t.m))^2Z_mfG.sq[k] <- (Z_mfG(t.f,t.m))^2claytons <- claytonTestSimulation(ped)clayton2df[k] <- claytons$Chi.squared.2.dfclayton1df[k] <- claytons$Chi.squared.1.dfplink.trend.f[k] <- (Z_fG)^2plink.geno.f[k] <- chisq.test(t.f)$statisticchi4df[k] <- chiComb(t.f, t.m)}# Combine the tests into a matrixtest_results <- cbind(Z_A.sq, Z_C.sq, Z_mfA.sq, Z_mfG.sq, clayton1df, clayton2df, plink.trend.f, plink.geno.f, chi4df)row.names(test_results) <- 1:n_repsreturn(test_results)}makeNullTable <- function(results, reps, alpha){        # Store the appropriate quantiles of the chi-squared distribution on 1 and 2 df        chi.1 <- qchisq(1 - alpha, 1)        chi.2 <- qchisq(1 - alpha, 2)        chi.4 <- qchisq(1 - alpha, 4)        prop.sig.tests <- matrix(rep(-100, reps*9), reps, 9)        colnames(prop.sig.tests) <- colnames(results)        prop.sig.tests[, c(1,3,4,5,7)] <- results[, c(1,3,4,5,7)] >= chi.1        prop.sig.tests[, c(2,6,8)] <- results[, c(2,6,8)] >= chi.2        prop.sig.tests[,9] <- results[,9] >= chi.4        # Calculate the proportion of times each test gives a significant result and the        appropriate standard error for inclusion in my paper        x <- colSums(prop.sig.tests, na.rm = TRUE)        n <- nrow(prop.sig.tests)        estAndCI <- binconf(x,n,alpha,"wilson")        # return_val is a vector of strings, with each element of the vector giving the estimated proportion of significant results followed by (1-alpha)% CI using the Wilson method. These values are rounded to 4 decimal places        return_val <- paste(sapply(round(estAndCI[,1],4), sprintf, fmt="%#.4f"), " (", sapply(round     (estAndCI[,2],4), sprintf, fmt="%#.4f"), " - ", sapply(round    (estAndCI[,3],4), sprintf, fmt="%#.4f"), ")", sep = "")        return(return_val)        }        makeTable <- function(results, reps, alpha){        # Store the appropriate quantiles of the chi-squared distribution on 1 and 2 df        chi.1 <- qchisq(1 - alpha, 1)        chi.2 <- qchisq(1 - alpha, 2)        chi.4 <- qchisq(1 - alpha, 4)        prop.sig.tests <- matrix(rep(-100, reps*9), reps, 9)        colnames(prop.sig.tests) <- colnames(results)        prop.sig.tests[, c(1,3,4,5,7)] <- results[, c(1,3,4,5,7)] >= chi.1        prop.sig.tests[, c(2,6,8)] <- results[, c(2,6,8)] >= chi.2        prop.sig.tests[,9] <- results[,9] >= chi.4        # Calculate the proportion of times each test gives a significant result and the        appropriate standard error for inclusion in my paper        p <- colMeans(prop.sig.tests, na.rm = TRUE)        return(p)        }